src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -442,37 +470,92 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
             return future;
         }
 
+        SpoolMessage spooledMessage;
         try {
-            spool.addMessage(request);
+            spooledMessage = spool.addMessage(request);
             triggerSpooler();
         } catch (InterruptedException | SpoolerStoreException e) {
             logger.atDebug().log("Fail to add publish request to spooler queue", e);
             future.completeExceptionally(e);
             return future;
         }
-        return CompletableFuture.completedFuture(0);
+        return spooledMessage.getFuture();
+    }
+
+    protected void isValidPublishRequest(PublishRequest request) throws SpoolerStoreException {
+        // Payload size should be smaller than MQTT maximum message size
+        int messageSize = request.getPayload().length;
+        if (messageSize > MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE) {
+            throw new SpoolerStoreException(String.format("The message size exceeds the maximum size of %d ",
+                    MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE));
+        }
+
+        // Topic should not contain wildcard characters
+        String topic = request.getTopic();
+        if (topic.contains("#") || topic.contains("+")) {
+            throw new SpoolerStoreException("The topic of publish request should not contain wildcard "
+                    + "characters of '#' or '+'");
+        }
+
+        String reservedTopicTemplate = "^\\$AWS/rules/\\S+/\\S+";
+        String prefixOfReservedTopic = "^\\$AWS/rules/\\S+?/";
+        if (Pattern.matches(reservedTopicTemplate, topic)) {
+            // remove the prefix of "$AWS/rules/rule-name/"
+            topic = topic.split(prefixOfReservedTopic, 2)[1];
+        }
+
+        // Topic should not have no more than maximum number of forward slashes (/)
+        if (topic.chars().filter(num -> num == '/').count() > DEFAULT_MAXIMUM_NUMBER_OF_FORWARD_SLASHES) {
+            logger.atInfo().kv("topic", topic).log("new topic - for /");
```

Comment:
this log seems really really weird. What are you trying to tell the user with this?

Reply:
Sorry, it is for testing. I will remove it here.